operator/controllers/keptnworkloadinstance/controller.go

```diff
@@ -91,43 +89,36 @@ func (r *KeptnWorkloadInstanceReconciler) Reconcile(ctx context.Context, req ctr
 		return reconcile.Result{}, fmt.Errorf("could not fetch KeptnWorkloadInstance: %+v", err)
 	}
 
-	r.Log.Info("Workload Instance found", "instance", workloadInstance)
-
-	if workloadInstance.IsPostDeploymentCompleted() {
-		return reconcile.Result{}, nil
+	if !workloadInstance.IsPreDeploymentCompleted() {
+		r.Log.Info("Pre deployment checks not finished")
+		err := r.reconcilePreDeployment(ctx, req, workloadInstance)
+		if err != nil {
+			r.Log.Error(err, "Error reconciling pre-deployment checks")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
 	}
 
-	r.Log.Info("Post deployment checks not finished")
-
-	isDeployed, err := r.UpdateWorkloadResourceDeploymentStatus(ctx, workloadInstance)
-	if err != nil {
-		return reconcile.Result{}, err
-	}
-	if isDeployed {
-		err = r.reconcilePostDeployment(ctx, req, workloadInstance)
+	if !workloadInstance.IsDeploymentCompleted() {
+		r.Log.Info("Deployment not finished")
+		err := r.reconcileDeployment(ctx, workloadInstance)
 		if err != nil {
-			r.Log.Error(err, "Error reconciling post-deployment checks")
+			r.Log.Error(err, "Error reconciling deployment")
 			return ctrl.Result{}, err
 		}
-		return ctrl.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+		return ctrl.Result{Requeue: true, RequeueAfter: 10 * time.Second}, nil
 	}
 
-	r.Log.Info("Deployment of pods not finished")
-
-	if workloadInstance.IsPreDeploymentCompleted() {
+	if !workloadInstance.IsPostDeploymentCompleted() {
 		r.Log.Info("Post deployment checks not finished")
+		err = r.reconcilePostDeployment(ctx, req, workloadInstance)
+		if err != nil {
+			r.Log.Error(err, "Error reconciling post-deployment checks")
+			return ctrl.Result{Requeue: true}, err
+		}
 		return ctrl.Result{Requeue: true}, nil
 	}
-
-	r.Log.Info("Pre deployment checks not finished")
-
-	err = r.reconcilePreDeployment(ctx, req, workloadInstance)
-	if err != nil {
-		r.Log.Error(err, "Error reconciling pre-deployment checks")
-		return ctrl.Result{}, err
-	}
-
-	return ctrl.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+	return ctrl.Result{Requeue: true, RequeueAfter: 30 * time.Second}, nil
```

Comment:
should we requeue at the end of this function? seems like we approach this line only if the post-deployments are already finished, so I assume we should not requeue again, what do you think ?

Reply:
Yes, we should requeue otherwise we have a single shot to check for pre-, post-, and the -deployment itself.  
The only time that we don't requeue is when the job is done at line 84.